### PR TITLE
Scan for selinux write error in registry diagnostic.

### PR DESCRIPTION
**Pull request by dgoodwin**
In the cluster admin diagnostic which scans the registry pod logs, add an
explicit check for an error which indicates the registry cannot write to disk,
which is a strong indicator for an selinux problem. The fix from the origin
troubleshooting page is then suggested.

We also watch for a successful write log entry, and if found *after* the error,
we know the problem has since been fixed and the diagnostic error can be safely
ignored.

This replaces pr #4926.

### CODE REVIEW

**Comment by dgoodwin** [comment]
Sample output:

ERROR: [DClu1020 from diagnostic ClusterRegistry@openshift/origin/pkg/diagnostics/cluster/registry.go:266]
       The pod logs for the "docker-registry-1-wlrhk" pod belonging to
       the "docker-registry" service indicated SELinux may be preventing
       the registry from writing to disk.
       
       This may be resolved on the node by running:
       
           sudo chcon -R -t svirt_sandbox_file_t [PATH_TO]/openshift.local.volumes
       
       time="2015-10-05T19:24:47Z" level=error msg="An error occured" err.code=UNKNOWN err.detail="mkdir /registry/docker/registry/v2/repositories/test/nodejs-ex/_uploads/5db6a12f-60dc-480a-8d37-84a2b3a6f8eb: permission denied" err.message="unknown error" http.request.host="172.30.150.249:5000" http.request.id=5362c6c0-235d-464e-a7ef-2b4eee2ac361 http.request.method=POST http.request.remoteaddr="172.17.42.1:54598" http.request.uri="/v2/test/nodejs-ex/blobs/uploads/" http.request.useragent="
docker/1.7.1 go/go1.4.2 kernel/3.10.0-229.el7.x86_64 os/linux arch/amd64" instance.id=767b7047-6624-4649-9fa6-18a4dc45f3cd vars.name="test/nodejs-ex"

**Comment by deads2k** [comment]
I didn't reason about correctness, but I like the approach.

@sosiouxme

**Comment by rhatdan** [comment]
Any reason you are not using :Z or :zap: 
Just chcon -t svirt_sandbox_file_t

Will allow all containers to write to this content (Equivalent of using :z in the volume command)

Using :Z will ensure the volume is private to the container.

**Comment by dgoodwin** [comment]
@rhatdan no reason! Just working off https://github.com/openshift/origin/blob/master/docs/debugging-openshift.md and my own lack of knowledge.

If it's ok I will update both places and include in this PR.

**Comment by dgoodwin** [comment]
@rhatdan could you clarify where/how I would use the :Z? I can't see any way to do so with chcon. Are you referring to this being a bug with how we execute docker run for the actual pod container?

**Comment by rhatdan** [comment]
docker run -v /src:/dst:Z ...

Is how you do this with docker.  I believe kubernetes has something the sends this command along too.  Private or Shared might be how it is specified in k8s.

**Comment by dgoodwin** [comment]
The actual params used to run the registry is unknown to me, I'd like to dig in and find out who/what launches that container with what params. If it's alright I'll file an issue and see if anyone responds with why that is the way it is before I get a chance to dig in.

WRT this diagnostic it's probably still good to get this out there as weird stuff happens, and people are definitely hitting this often.

Test failures do not look related to the PR itself as far as I can tell.

@sosiouxme does this diagnostic look healthy to you?

**Comment by sosiouxme** [comment]
@dgoodwin LGTM

**Comment by dgoodwin** [comment]
Bump! Anything left to do for a merge @deads2k ?

File: pkg/diagnostics/cluster/registry.go [hunk 1]
```diff
@@ -221,8 +233,15 @@ func (d *ClusterRegistry) checkRegistryLogs(pod *kapi.Pod, r types.DiagnosticRes
 	}
 	defer readCloser.Close()
 
+	// Indicator that selinux is blocking the registry from writing to disk:
+	selinuxErrorRegex, _ := regexp.Compile(".*level=error.*mkdir.*permission denied.*")
```
Comment (deads2k) [inline pkg/diagnostics/cluster/registry.go]: This is the same message you get if your volume doesn't allow writing by any user, right?  If so, update the help message to note both problems.

**Comment by deads2k** [comment]
> Bump! Anything left to do for a merge @deads2k ?

Just one question/comment.

**Comment by dgoodwin** [comment]
Updated to mention ownership/permissions problems and a suggestion to consult the persistent storage section of admin guide.

**Comment by deads2k** [comment]
[merge]

@sosiouxme Diagnostics like this are going to be fragile.  Useful, but fragile.  Do you have a card to create extended tests to make sure they don't rot?

**Pull request merged**